### PR TITLE
Add simple symbol table to QASM 3 exporter

### LINKED ISSUE #7742: OpenQASM3 exporter emits variables with reserved keywords
### Environment

- **Qiskit Terra version**: 0.19.2
- **Python version**: 3.8 
- **Operating system**:  OSX

### What is happening?

When dumping QASM3 from a Qiskit circuit the emitted QASM3 has variables with names that are reserved keywords.

### How can we reproduce the issue?

```python
from qiskit import IBMQ, transpile, qasm3
from qiskit_experiments.library import QuantumVolume

from qiskit.test.mock import FakeMontreal

qv_exp = QuantumVolume([0, 1], seed=42)
circuits = qv_exp.circuits()

backend = FakeMontreal()

def dump_qasm3(circuit, backend=backend):
    return qasm3.Exporter(includes=[], basis_gates=backend.configuration().basis_gates).dumps(circuit)

transpiled_circuits = transpile(circuits, backend, optimization_level=0)
print(dump_qasm3(transpiled_circuits[0]))
```
Emits:
```qasm
OPENQASM 3;
bit[2] measure;
rz(-0.527043787493904) $0;
sx $0;
rz(5.16854203223675) $0;
sx $0;
rz(7.52041849263069) $0;
rz(2.52883547684609) $1;
sx $1;
rz(4.18263287241389) $1;
sx $1;
rz(11.4206034322753) $1;
cx $0, $1;
rz(pi/2) $0;
sx $0;
rz(3.83403185518176) $0;
sx $0;
rz(5*pi/2) $0;
rz(-3.13722964498385) $1;
sx $1;
rz(3.43607384014569) $1;
sx $1;
rz(7.84942236436758) $1;
cx $0, $1;
rz(pi/2) $0;
sx $0;
rz(3.15129722109744) $0;
sx $0;
rz(2*pi) $0;
rz(-pi) $1;
sx $1;
rz(4.71106568778747) $1;
sx $1;
rz(10.9942509926498) $1;
cx $0, $1;
rz(0.0407864518524126) $0;
sx $0;
rz(3.68646663680408) $0;
sx $0;
rz(8.261303299549) $0;
rz(0.384886550311654) $0;
sx $0;
rz(4.72101136719892) $0;
sx $0;
rz(7.05118934560937) $0;
rz(0.360649613151623) $1;
sx $1;
rz(4.92286155451527) $1;
sx $1;
rz(11.7465986327618) $1;
rz(-0.879779034806911) $1;
sx $1;
rz(5.92954645622555) $1;
sx $1;
rz(7.17120533229796) $1;
cx $0, $1;
rz(pi/2) $0;
sx $0;
rz(4.06066473777189) $0;
sx $0;
rz(5*pi/2) $0;
rz(-3.1391797270208) $1;
sx $1;
rz(3.64322156481995) $1;
sx $1;
rz(7.85122966784221) $1;
cx $0, $1;
rz(-pi/2) $0;
sx $0;
rz(3.7166355179822) $0;
sx $0;
rz(3*pi) $0;
rz(-pi) $1;
sx $1;
rz(4.71106568778747) $1;
sx $1;
rz(10.9942509926498) $1;
cx $0, $1;
rz(-1.47560665166994) $0;
sx $0;
rz(4.18505473328482) $0;
sx $0;
rz(9.47043569736961) $0;
rz(0.398413507779781) $1;
sx $1;
rz(4.1804199287629) $1;
sx $1;
rz(12.3633552711276) $1;
barrier $0, $1;
measure[0] = measure $0;
measure[1] = measure $1;
```
Note the `bit[2] measure;` declaration which is invalid.

### What should happen?

The QASM3 exporter should mangle the names of reserved keywords for classical/qubit registers with names that collied.

### Any suggestions?

_No response_

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

This is a very simplistic implementation of a symbol table for the
OpenQASM 3 exporter, which allows us to rebind the names of register and
parameter names when they can't be used verbatim.  This is necessary to
fix a couple of bugs:

  - variables that shared a name with a reserved keyword would produce
    invalid QASM, because the name would be output verbatim

  - registers and parameters could have the same name since they are not
    stored in the same table within `QuantumCircuit`.

The scoping and shadowing rules are not properly defined in the
specification yet, and we expect to completely overhaul how expressions
are stored in Terra in the near future, which may prompt a much larger
overhaul of how the exporter works.

### Details and comments

Fix #7742.

### CODE REVIEW

**Comment by jakelishman** [comment]
@1ucian0 or @kevinhartman: if one of you have time to have a flick through this in the next few days (no huge rush, but it'd be good for Terra 0.20) that'd be great, thank you.

File: releasenotes/notes/qasm3-escape-reserved-keywords-60d463db36d96319.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+---
+fixes:
+  - |
+    The OpenQASM 3 exporter will now escape register and parameter names that
+    clash with reserved OpenQASM 3 keywords by generating a new unique name.
+  - Registers and parameters with the same name will no longer have naming
```
Comment (1ucian0) [inline releasenotes/notes/qasm3-escape-reserved-keywords-60d463db36d96319.yaml:6]:
```suggestion
    Registers and parameters with the same name will no longer have naming
```

Comment (1ucian0) [inline releasenotes/notes/qasm3-escape-reserved-keywords-60d463db36d96319.yaml:6]:
```suggestion
    Registers and parameters with the same name will no longer have naming
```

**Review by 1ucian0 — DISMISSED** [review]
Thanks!

**Review by mtreinish — APPROVED** [review]

**Comment by jakelishman** [comment]
Merged manually because the mergify bot got confused after the `black` CI failures last night - all required tests had passed.

**Pull request merged**